llvm/include/llvm/CodeGen/SDPatternMatch.h

```diff
@@ -457,6 +457,53 @@ struct BinaryOpc_match {
   }
 };
 
+template <typename LHS_t, typename RHS_t, typename PredFuncT>
+struct AnyBinaryOp_match {
+  LHS_t L;
+  RHS_t R;
+  PredFuncT PredFunc;
+  bool Commutable;
+
+  AnyBinaryOp_match(const PredFuncT &Pred, const LHS_t &LHS, const RHS_t &RHS,
+                    const bool Commutable)
+      : PredFunc(Pred), L(LHS), R(RHS), Commutable(Commutable) {}
+
+  template <typename OpTy, typename MatchContext>
+  bool match(OpTy *V, const MatchContext &Ctx) {
+    assert(Ctx.getTLI() && "TargetLowering is required for this pattern");
+    if (auto *I = dyn_cast<BinaryOperator>(V))
+      return (PredFunc(*Ctx.getTLI()) &&
+              ((L.match(I->getOperand(0)) && R.match(I->getOperand(1))) ||
+               (Commutable && L.match(I->getOperand(1)) &&
+                R.match(I->getOperand(0)))));
+    return false;
+  }
+};
+
+template <typename LHS, typename RHS>
+inline auto m_AnyBinOp(const LHS &L, const RHS &R) {
+  return AnyBinaryOp_match{[](const TargetLowering &TLI) { return true; }, L, R,
+                           false};
+}
+template <typename LHS, typename RHS>
+inline auto m_c_AnyBinOp(const LHS &L, const RHS &R) {
+  return AnyBinaryOp_match{[](const TargetLowering &TLI) { return true; }, L, R,
+                           true};
+}
+
+template <typename LHS, typename RHS>
+inline auto m_AnyBinOp(unsigned Opc, const LHS &L, const RHS &R) {
```

Comment:
This is wrong, we want to extract the opcode in the match

Reply:
done.